keysync_responder.go

```diff
@@ -0,0 +1,104 @@
+package nitriding
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hf/nitrite"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+var (
+	errFailedNonce     = "failed to create nonce"
+	errNoBase64        = "failed to Base64-decode attestation document"
+	errFailedVerify    = "failed to verify attestation document"
+	errFailedFindNonce = "could not find provided nonce"
+	errInvalidSbKeys   = "invalid secretbox key material"
+)
+
+type timeFunc func() time.Time
+
+// getNonceHandler returns a HandlerFunc that creates a new nonce and returns
+// it to the client.
+func getNonceHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		nonce, err := newNonce()
+		if err != nil {
+			http.Error(w, errFailedNonce, http.StatusInternalServerError)
+			return
+		}
+
+		e.nonceCache.Add(nonce.B64())
+		fmt.Fprintf(w, "%s\n", nonce.B64())
+	}
+}
+
+func getKeysHandler(e *Enclave, curTime timeFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var ourNonce, theirNonce nonce
+
+		body, err := ioutil.ReadAll(r.Body)
```

Comment:
Can this be limited to `maxAttDocLen` as well?

Reply:
Fixed in de103f3, after an earlier, failed attempt. This was surprisingly annoying to address, and required a custom Reader implementation.